### PR TITLE
Add `&pin (mut|const) T` type position sugar

### PR DESCRIPTION
**Pull request by eholk**
This adds parser support for `&pin mut T` and `&pin const T` references. These are desugared to `Pin<&mut T>` and `Pin<&T>` in the AST lowering phases.

This PR currently includes #130526 since that one is in the commit queue. Only the most recent commits (bd450027eb4a94b814a7dd9c0fa29102e6361149 and following) are new.

Tracking:

- #130494 

r? @compiler-errors

### CODE REVIEW

File: compiler/rustc_ast/src/util/classify.rs
Comment (compiler-errors) [inline compiler/rustc_ast/src/util/classify.rs:250]: I'd strongly prefer if we split this out into a totally different `TyKind`. These are not refs, and I think the last time we've shoved new options into a preexisting type (`dyn*`) it led to more headaches than expected.

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_ast/src/util/classify.rs
Comment (eholk) [reply compiler/rustc_ast/src/util/classify.rs:250]: That makes good sense to me. I'll work on that.

**Comment by eholk** [comment]
I'm missing a test to make sure that `&pin i32` _doesn't_ parse. I'm also failing that test, so let me add that and fix it.

@rustbot author

File: compiler/rustc_parse/src/parser/ty.rs
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/ty.rs:593]: This should get fixed before this lands.

Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/ty.rs:593]: Also, please in the future associate any fixmes with the feature. `FIXME(pin_ergonomics)`. It makes it easier to track.

File: compiler/rustc_ast/src/util/classify.rs
Comment (eholk) [reply compiler/rustc_ast/src/util/classify.rs]: Okay, I've broken out `PinnedRef` as a separate variant. In most cases I'm still treating them the same, but that's more obvious in the source now.

File: compiler/rustc_parse/src/parser/ty.rs
Comment (eholk) [reply compiler/rustc_parse/src/parser/ty.rs]: This is fixed now, along with tests to catch some of the more interesting subtleties here.

File: src/tools/rustfmt/src/types.rs
Comment (compiler-errors) [inline src/tools/rustfmt/src/types.rs:734]: Why not just fix it rather than suppress it?

Comment (compiler-errors) [reply src/tools/rustfmt/src/types.rs:734]: The nightly style policy states that you can implement formatting, for the record.

**Comment by compiler-errors** [comment]
@rustbot author

Remember to switch back to S-waiting-on-review when it's rebased and updated

File: compiler/rustc_ast/src/ast.rs
Comment (eholk) [reply compiler/rustc_ast/src/ast.rs:2164]: Good idea. Done!

File: src/tools/rustfmt/src/types.rs
Comment (eholk) [reply src/tools/rustfmt/src/types.rs]: I was just being lazy. I added the basic support now, although I'm sure it will need more fine tuning to break lines in the right place and all that.

File: tests/ui/feature-gates/feature-gate-pin_ergonomics.rs
Comment (oli-obk) [inline tests/ui/feature-gates/feature-gate-pin_ergonomics.rs:14]: What will happen if we had a macro on stable that has a type fragment in the first macro matcher and a token tree in the second. Would we now pick the first and error or sth if used as `themacro!(&pin const T)`

Comment (compiler-errors) [reply tests/ui/feature-gates/feature-gate-pin_ergonomics.rs:14]: Arbitrary token trees (nor `mut` nor `const`) are in the follow set for a type fragment, so this shouldn't happen afaict?

Comment (oli-obk) [reply tests/ui/feature-gates/feature-gate-pin_ergonomics.rs:14]:
```rust
macro_rules! foo {
    ($x:ty) => { compile_error!("bad") };
    ($($t:tt)*) => {}
}

foo!(&pin mut u32);
```

Compiles today, but may be affected by this PR either directly or just if the feature is enabled

Comment (compiler-errors) [reply tests/ui/feature-gates/feature-gate-pin_ergonomics.rs:14]: Yeah, though I'd consider this to be a theoretical breakage, similarly to adding `async` and `const` to trait bounds.

Comment (oli-obk) [reply tests/ui/feature-gates/feature-gate-pin_ergonomics.rs:14]: Agreed, but it should probably not break while the feature is still unstable I think. Only at stabilization time

Comment (compiler-errors) [reply tests/ui/feature-gates/feature-gate-pin_ergonomics.rs:14]: I'm not certain if there's a feasible way to prevent this breakage without simply excluding `&pin mut T` from the type nonterminal, which seems not good

Comment (eholk) [reply tests/ui/feature-gates/feature-gate-pin_ergonomics.rs:14]: I think this is covered by [RFC 3531](https://rust-lang.github.io/rfcs/3531-macro-fragment-policy.html). Basically, until the edition after this is hypothetically stabilized, `$ty` continues to match what it does currently. Then for the next edition, we'd update `$ty` to include the new syntax and add `$ty_2024` or some other more descriptive name to keep the existing behavior.

Comment (oli-obk) [reply tests/ui/feature-gates/feature-gate-pin_ergonomics.rs:14]: That's a lot of complexity. I agree that this is a highly artificial situation and will not occur in practice. If it had been easy to avoid we could have done so, but it isn't, so let's go ahead without any additional checks

Comment (eholk) [reply tests/ui/feature-gates/feature-gate-pin_ergonomics.rs:14]: Do you think it's worth adding your example as a test case to make sure we don't accidentally change the behavior of the `ty` matcher? I'd be happy to add it, although I think the matcher implementation is surprisingly separate from the parser so I'm not sure there's a huge risk of accidentally doing this.

Comment (oli-obk) [reply tests/ui/feature-gates/feature-gate-pin_ergonomics.rs:14]: I don't think it's needed, but don't feel strongly either way.

Someone from @rust-lang/lang should probably look at this and decide if it's worthwile discussing or even FCPing.

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated +I-lang-easy-decision

In Rust 1.41, we broke this program...

```rust
macro_rules! m {
    ($x:expr) => { compile_error!("bad") };
    ($($t:tt)*) => {}
}
m!(&raw const ());
```

...even though we didn't end up stabilizing `&raw const ..` until Rust 1.82.

Then we broke this in Rust 1.77:

```rust
macro_rules! m {
    ($x:ty) => { compile_error!("bad") };
    ($($t:tt)*) => {}
}
m!(impl const Sized);
```

(We haven't stabilized `const` trait modifiers yet.)  And then in Rust 1.78, we broke:

```rust
macro_rules! m {
    ($x:ty) => { compile_error!("bad") };
    ($($t:tt)*) => {}
}
m!(impl async Sized);
```

(We haven't stabilized `async` trait modifiers yet.)

Now we plan to break this program when this PR lands:

```rust
macro_rules! m {
    ($x:ty) => { compile_error!("bad") };
    ($($t:tt)*) => {}
}
m!(&pin const ());
```

This breakage is believed to be theoretical-only.  Does this sound OK to us?

(See [this thread](https://github.com/rust-lang/rust/pull/130635#discussion_r1774831501) for background.)

**Comment by compiler-errors** [comment]
The breakage specifically represents an inherent limitation to the "macro follow-set" formulation which is *supposed* to make us more resilient against extensions to the grammar like this.

Given two macro matcher arms:
* `$ty:ty`
* `($tt:tt)*`

And given tokens like:

* `&` `pin` `mut` [...more tokens may follow...]

On nightly today, `&pin` gets parsed as a type. However, we run out of matchers but still have tokens left (the `mut` token is next), so we fall through to the next arm. Since it's written like `($tt:tt)*`, everything is allowed, and we match the second arm successfully...

I think that's weird, because if this second arm were written like `$ty:ty mut`, that would be illegal, since `mut` is not in the follow-set of the `:ty` matcher. Thus, we can use `:tt` matchers to observe whether the compiler *actually* parses things not in our grammar that should otherwise be protected against, which seems pretty gross. 

I think this breakage is purely theoretical, though. I would like to hear any suggestions to close this hole and make macros a bit more resilient, but nothing really comes to mind :(

**Comment by eholk** [comment]
Given that this changes behavior even without the `pin_ergonomics` feature flag, I'd like to take a little time to find a workaround. I'm not particularly optimistic since this has been challenging in the past, but I'd like to at least take a look.

@rustbot author

**Comment by tmandry** [comment]
We are going to need `pin` in expressions, too, right?

Today we have `let x = pin!(foo);`
Tomorrow we might want `let x = &pin mut foo;`

Looks like you are relying on `&pin const` and `&pin mut` to disambiguate, which worked for raw references so I guess it can work here.

We can keep using this or the pin macro until the next edition and reserve a keyword then (but we'll have to figure out what to do about `Box::pin`).

**Comment by traviscross** [comment]
> We are going to need pin in expressions, too, right?

Yes, so we'd also be intending to break:

```rust
//@ check-pass
macro_rules! m {
    ($x:expr) => { compile_error!("bad") };
    ($($t:tt)*) => {}
}
m!(&pin const ());
```

**Comment by Noratrieb** [comment]
> I would like to hear any suggestions to close this hole and make macros a bit more resilient, but nothing really comes to mind :(

a bit off-topic for the thread but I believe a solution to this would be the following new logic:
- after the end of a macro matcher arm has been reached
- and there are still input tokens remaining
- and if the last part of the matcher is a metavar
- ensure that the first remaining token is in the follow set of this metavar
- if it is, move on to the next arm
- if it is not, **emit an error**

What this semantically does is strengthen the "commit to fully matching metavars or error" behavior such that it extends past the end.
I don't know how many macros rely on this, but it seems like emitting an FCW (instead of error) on such macro invocations would find all these cases and ensure that the follow-set logic is actually robust past the end.
But imo this shouldn't block this PR (which should probably just ship as-is) and can be done separately.

**Comment by traviscross** [comment]
That seems plausible to me.  I've filed and nominated...

- https://github.com/rust-lang/rust/issues/131025

...for the proposal for that FCW.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated +T-lang +relnotes

We discussed this in the meeting today.  We agreed that this seems like theoretical-only breakage and is OK to move forward as far as lang is concerned.  We'll mention this in the release notes as it is technically breaking, and in that way we'll help to make this information available to those doing the beta crater runs, just in case anything is found there.

Our general sentiment was that anyone using match arms as an oracle to detect what gets parsed in other arms is probably asking for trouble; it's unwise to rely on the subtleties of this, given the limitations of the follow-set behavior with respect to the interaction between arms, and we're interested in better ways to explicitly reserve this space.

**Comment by eholk** [comment]
Based on the lang team discussion, I think we should move forward on this PR and look at solving the problem more generally separately (perhaps with #131025).

@rustbot ready

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**
